Ensure overlaps_with_timerange and intersect_with are consistent

philipnbbc opened the pull request:
# Details
It makes sense to view an overlap occurring if the intersection is not empty.

The same should also happen to the special case of `never()` and `eternity()` timeranges. It should return False given that `intersect_with` returns an empty (`never()`) timerange.

This also resulted in fixes to the tests for `is_contiguous_with_timerange` where a timerange is never().

# Pivotal Story
Story URL: https://www.pivotaltracker.com/story/show/188190520

# Related PRs
_Where appropriate. Indicate order to be merged._

# Links to external test runs/working deployment
_Where appropriate, if separate to default CI run_

# Submitter PR Checks
_(tick as appropriate)_

- [ ] PR completes task/fixes bug
- [ ] Tests exercise code appropriately
- [ ] New features and API breaks are flagged in commit messages using magic strings
- [ ] Documentation updated (README, Confluence, Docstrings, API spec, Engineering Guide, etc.)
- [ ] Downstream repos have been checked for potential breaks & fixed as needed
- [ ] APIs/UIs/CLIs updated as required
- [ ] PR added to Pivotal story
- [ ] Follow-up stories added to Pivotal
- [ ] Any temporary code/configuration removed (e.g. test deployment environment, temporary commontooling branch)
- [ ] Any pins against pre-releases have been removed

# Reviewer PR Checks
_(tick as appropriate)_

- [ ] PR completes task/fixes bug
- [ ] Tests exercise code appropriately
- [ ] Design makes sense, and fits with our current code base
- [ ] Code is easy to follow
- [ ] PR size is sensible
- [ ] Commit history is sensible and tidy

# Info on Cloudfit PRs
- The checks above are guidelines. They don't all have to be ticked, but they should all have been considered.
- For more details on how to asses PRs, see: https://github.com/bbc/rd-apmm-docs-ways-of-working/blob/master/workflow/PRChecklist.md

samdbmg reviewed (approved):
This LGTM.

Two questions:
1. Should this have a sem-ver comment? It is a bug, but one that changes a long-standing behaviour, so maybe at least `sem-ver: feature`? I'm on the fence on this one
2. Do `TimeValueRange` and `CountRange` need the same fix?

philipnbbc commented:
> Two questions:
> 
>     1. Should this have a sem-ver comment? It is a bug, but one that changes a long-standing behaviour, so maybe at least `sem-ver: feature`? I'm on the fence on this one
> 
>     2. Do `TimeValueRange` and `CountRange` need the same fix?

1. Changed to a semver feature. I checked our code whether this behaviour was used (it wasn't), but there could be users elsewhere that we don't know about.
2. Thanks for spotting. Fixed.
